--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-rk.c	2025-06-04 16:46:27.000000000 +0400
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-rk.c	2025-06-08 09:46:25.991903867 +0400
@@ -1595,12 +1595,21 @@
 		clk_set_rate(bsp_priv->clk_mac, 50000000);
 	}
 
-	if (plat->phy_node && bsp_priv->integrated_phy) {
+	if (plat->phy_node) {
 		bsp_priv->clk_phy = of_clk_get(plat->phy_node, 0);
 		ret = PTR_ERR_OR_ZERO(bsp_priv->clk_phy);
 		if (ret)
 			return dev_err_probe(dev, ret, "Cannot get PHY clock\n");
 		clk_set_rate(bsp_priv->clk_phy, 50000000);
+		/* If it is not integrated_phy, clk_phy is optional */
+		if (bsp_priv->integrated_phy) {
+			if (IS_ERR(bsp_priv->clk_phy)) {
+				ret = PTR_ERR(bsp_priv->clk_phy);
+				dev_err(dev, "Cannot get PHY clock: %d\n", ret);
+				return -EINVAL;
+			}
+			clk_set_rate(bsp_priv->clk_phy, 50000000);
+		}
 	}
 
 	return 0;
@@ -1612,6 +1621,8 @@
 
 	if (enable) {
 		if (!bsp_priv->clk_enabled) {
+			if (!IS_ERR(bsp_priv->clk_phy))
+				clk_prepare_enable(bsp_priv->clk_phy);
 			ret = clk_bulk_prepare_enable(bsp_priv->num_clks,
 						      bsp_priv->clks);
 			if (ret)
